src/Database/Relations/HasOneOrMany.php

```diff
@@ -56,6 +56,21 @@ public function create(array $attributes = [], $sessionKey = null)
     public function add(Model $model, $sessionKey = null)
     {
         if ($sessionKey === null) {
+            /**
+             * @event model.relation.beforeAdd
+             * Called before adding a relation to the model (for AttachOneOrMany, HasOneOrMany & MorphOneOrMany relations)
+             *
+             * Example usage:
+             *
+             *     $model->bindEvent('model.relation.beforeAdd', function (string $relationName, \October\Rain\Database\Model $relatedModel) use (\October\Rain\Database\Model $model) {
+             *         if ($relationName === 'dummyRelation') {
+             *             throw new \Exception("Invalid relation!");
+             *         }
+             *     });
+             *
+             */
+            $this->parent->fireEvent('model.relation.beforeAdd', [$this->relationName, $this->related]);
```

Comment:
Identically as above. Related model here is method's `$model`. Identically in _afterAdd_, _before/afterRemove_ and _MorphOneOrMany_ too.